dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/relation/RelationModelManager.java

```diff
@@ -0,0 +1,596 @@
+        }
+    }
+
+    // Getting the correct relation to extract is tricky.
+    // In the case of redefinition, we have to find the latest defined one which we care about only.
+    // If there is no redefinition the original one will be returned simply.
+    private Relation getRelationWithName(String name) {
+        final Map<Integer, Relation> redefineds = new HashMap<>();
+        for (Relation r : wmm.getRelations()) {
+            final int defIndex = r.getNames().stream().filter(s -> s.startsWith(name + "#"))
+                                  .map(s -> {
+                                    try {
+                                        return Integer.parseInt(s.substring(s.lastIndexOf("#") + 1));
+                                    } catch (NumberFormatException e) {
+                                        return Integer.MIN_VALUE;
+                                    }
+                                  })
+                                  .filter(i -> i != Integer.MIN_VALUE)
+                                  .max(Comparator.naturalOrder()).orElse(0);
+            if (defIndex != 0) {
+                redefineds.put(defIndex, r);
+            }
+        }
+        return redefineds.entrySet().stream().max(Map.Entry.comparingByKey())
+                         .map(Map.Entry::getValue).orElse(wmm.getRelation(name));
+    }
+
+    private void extractRelations(List<String> relationNames) {
+        Set<Relation> relsToExtract = new HashSet<>();
+        for (String name : relationNames) {
+            Relation r = getRelationWithName(name);
+            if (r == null) {
+                logger.warn("Relation with the name {} does not exist", name);
+                continue;
+            }
+            relsToExtract.add(r);
+            createModel(r, name);
+        }
+
+        Set<RelationGraph> relGraphs = new HashSet<>();
+        DependencyGraph<Relation> dependencyGraph = DependencyGraph.from(relsToExtract);
+
+        for (Set<DependencyGraph<Relation>.Node> component : dependencyGraph.getSCCs()) {
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = new RecursiveGraph();
+                    recGraph.setName(r.getNameOrTerm() + "_rec");
+                    relGraphs.add(recGraph);
+                    relGraphCache.put(r, recGraph);
+                }
+            }
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                RelationGraph rg = createGraph(r);
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = (RecursiveGraph) relGraphCache.get(r);
+                    recGraph.setConcreteGraph(rg);
+                } else {
+                    relGraphs.add(rg);
+                }
+            }
+        }
```

Comment:
is it necessary to first process all recursive relations? I.e., do we need to have the two loops?

Reply:
Yes, because unlike `Wmm`, the CAAT solver uses special recursive placeholder relations (like we used to have back in the days). Those placeholders might not be strictly necessary, but they make computation easier.